### PR TITLE
Throw error when ThermalContact is used with uniform mesh refinement

### LINKED ISSUE #13043: ThermalContact does not work with -r 1 uniform mesh refinement
## Rationale

It appears that `ThermalContact` does not work with uniform mesh refinement. 

## Description

No heat transfer across the gap if uniform mesh refinement is specified. Example:
```
/moose/modules/heat_conduction/test/tests/heat_conduction/2d_quadrature_gap_heat_transfer/gap_conductivity_property.i
```

Execute using provided mesh:
```
../../../../heat_conduction-opt -i gap_conductivity_property.i
```
Report of average block temperatures, `temp_1` and `temp_2`:
```
+----------------+----------------+----------------+
| time           | temp_1         | temp_2         |
+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|   1.000000e+00 |   3.214286e+02 |   3.785714e+02 |
+----------------+----------------+----------------+
```

Execute with `-r 1`:
```
../../../../heat_conduction-opt -i gap_conductivity_property.i -r 1
```
Average block temperatures:
```
+----------------+----------------+----------------+
| time           | temp_1         | temp_2         |
+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|   1.000000e+00 |   3.000000e+02 |   4.000000e+02 |
+----------------+----------------+----------------+
```

These are the left and right boundary conditions, respectively, and hence the
solution you'd expect when you insulate the two blocks from each other. 

## Impact

Bug fix.

### PR DESCRIPTION
**Pull request by snschune**
closes #13043

### CODE REVIEW

File: modules/heat_conduction/src/bcs/GapHeatTransfer.C [hunk 1]
```diff
@@ -194,8 +190,9 @@ GapHeatTransfer::computeQpJacobian()
   if (!_has_info)
     return 0.0;
 
-  return _test[_i][_qp] * ((_u[_qp] - _gap_temp) * _edge_multiplier * _gap_conductance_dT[_qp] +
-                           _edge_multiplier * _gap_conductance[_qp]) *
+  return _test[_i][_qp] *
+         ((_u[_qp] - _gap_temp) * _edge_multiplier * _gap_conductance_dT[_qp] +
+          _edge_multiplier * _gap_conductance[_qp]) *
```
Comment (tophmatthews) [inline modules/heat_conduction/src/bcs/GapHeatTransfer.C]: While you're in here...shouldn't _edge_multiplier be multiplied only once here? Residual is
```
(_u[_qp] - _gap_temp) * _edge_multiplier * _gap_conductance[_qp]
```

Comment (tophmatthews) [reply modules/heat_conduction/src/bcs/GapHeatTransfer.C]: wait scratch that...I didn't see the +

File: framework/src/auxkernels/GapValueAux.C [hunk 2]
```diff
@@ -104,16 +101,14 @@ GapValueAux::computeValue()
...
-      mooseWarning(msg.str());
-    }
-  }
+    mooseDoOnce(mooseWarning("No gap value information found for node ",
+                             current_node->id(),
+                             " on processor ",
+                             processor_id(),
+                             " at coordinate ",
+                             Point(*current_node),
+                             ". The computed result will most likely be wrong.\n A likely cause is "
```
Comment (bwspenc) [inline framework/src/auxkernels/GapValueAux.C]: There are lots of cases where you might not project to another surface. This message would be misleading.

**Review by bwspenc — COMMENTED** [review]
This will generate warnings for many legitimate cases in which a part of a surface is simply not in contact.  This seems like a poor solution for the problem of missing nodes with mesh refinement. I know a while back I did something to reinitialize the contact search for mechanical contact to fix a similar issue. We should do something like that to fix this the right way instead.

**Comment by snschune** [comment]
@bwspenc If there are legitimate cases where `pinfo` should be `NULL`, then I agree this is not a good solution. Something needs to be done with this though, because it can seriously mess up casual/first time users of thermal contact. 

My case was incredibly simple, two non-moving concentric cylinders with a filler gas in between. The setup of the classes require you to go all in with thermal contact even if you only want heat transfer across a fixed gap - which is a legitimate need for many applications. 

This is important for high temperature reactor applications so I would be happy to be involved in making this more robust and user friendly. 

For now, what about checking against `MooseMesh`'s `uniform_refine` level and throwing and error 
if it's not zero.

**Comment by bwspenc** [comment]
@snschune Sorry for not being very responsive. I've been out for a while but just got back today. Yes, there are a lot of legitimate cases in which a node doesn't project to a surface. Just about every one of our Bison models does that.

I'm fine with your proposed fix that just errors out if you are using refinement together with thermal contact as a temporary solution.

I'm kind of curious about why this doesn't work with thermal contact because I know we resolved the same issue with mechanical contact, and it kind of seems like the fix that worked there should have fixed the problem for thermal contact as well. In the case you ran that caused trouble, were you using the `quadrature` option?

**Comment by snschune** [comment]
I tested both options and neither one worked if I recall correctly.

**Comment by snschune** [comment]
@bwspenc regardless of `quadrature` option, the answer with uniform refinement is wrong.

**Comment by bwspenc** [comment]
I like your new proposed check.

**Review by bwspenc — DISMISSED** [review]
I like this approach to handling the problem. I'd like to hear the opinion of @permcody or @friedmud. Could you add a test of this to verify that we hit this error message?  That problem would also be useful for verifying that thermal contact does work with refinement once we get that worked out.

**Review by permcody — APPROVED** [review]
This seems like a reasonable workaround in the short term. Thanks!

**Pull request merged**